backend/srs/converters/StructSerializer.cpp

```diff
@@ -1,13 +1,108 @@
+            marker_data_compact.timestamp_low_bits =
+                static_cast<decltype(marker_data_compact.timestamp_low_bits)>(timestamp_low_bits.to_ulong());
+            marker_data_compact.timestamp_high_bits =
+                static_cast<decltype(marker_data_compact.timestamp_high_bits)>(timestamp_high_bits.to_ulong());
+            marker_data_compact.flag = static_cast<decltype(marker_data_compact.flag)>(0);
+            marker_data_compact.vmm_id = static_cast<decltype(marker_data_compact.vmm_id)>(marker_data.vmm_id);
+            return {};
+        }
+
+        auto hit_to_compact(const HitData& hit_data, internal::HitDataCompact& hit_data_compact)
+            -> std::expected<void, std::string_view>
+        {
+            auto is_correct_sizes =
+                check_compact_max_size(hit_data.channel_num, internal::CHANNEL_NUM_BIT_LENGTH)
+                    .and_then([&] { return check_compact_max_size(hit_data.bc_id, internal::BC_ID_BIT_LENGTH); })
+                    .and_then([&] { return check_compact_max_size(hit_data.adc, internal::ADC_BIT_LENGTH); })
+                    .and_then([&] { return check_compact_max_size(hit_data.vmm_id, internal::VMM_ID_BIT_LENGTH); })
+                    .and_then([&] { return check_compact_max_size(hit_data.offset, internal::OFFSET_BIT_LENGTH); });
+            if (!is_correct_sizes)
+            {
+                return std::unexpected(is_correct_sizes.error());
+            }
+            hit_data_compact.vmm_id = static_cast<decltype(hit_data_compact.vmm_id)>(hit_data.vmm_id);
+            hit_data_compact.flag = static_cast<decltype(hit_data_compact.flag)>(1);
+            hit_data_compact.adc = static_cast<decltype(hit_data_compact.adc)>(hit_data.adc);
+            hit_data_compact.bc_id = static_cast<decltype(hit_data_compact.bc_id)>(hit_data.bc_id);
+            hit_data_compact.bc_id = common::binary_to_gray(hit_data_compact.bc_id);
+            hit_data_compact.channel_num = static_cast<decltype(hit_data_compact.channel_num)>(hit_data.channel_num);
+            hit_data_compact.is_over_threshold =
+                static_cast<decltype(hit_data_compact.is_over_threshold)>(hit_data.is_over_threshold);
+            hit_data_compact.offset = static_cast<decltype(hit_data_compact.offset)>(hit_data.offset);
+            return {};
+        }
+
+        template <class T>
+            requires(std::same_as<std::remove_cvref_t<T>, internal::HitDataCompact> ||
+                     std::same_as<std::remove_cvref_t<T>, internal::MarkerDataCompact>)
+        auto compact_to_vector(const T& compact_data) -> std::vector<char>
+        {
+            auto compact_bitset =
+                std::bitset<sizeof(std::uint64_t) * common::BYTE_BIT_LENGTH>{ std::bit_cast<uint64_t>(compact_data) };
+            auto output = std::vector<char>{};
+            output.resize(sizeof(std::uint64_t) / common::BYTE_BIT_LENGTH);
```

Comment:
Well, this is not a good decision to create a new vector in a loop. When a new vector is created, you have the memory allocation (in heap), which is very slow and should be avoided.

What you could do is to create some buffer vector and initialize it only once in the constructor. In the loop, before you write something into it, reset all its members to be zero. Your `compact_to_vector` should take an non-const ref as its output:

```cpp
void compact (const T& compact_data, std::vector<char>& output);
```